File: JavaScript/parse.js
import { tokenize } from "./lex.js";
function parse(code) {
  const result = [];
  const tokens = tokenize(code);
  while (tokens.length) {
    result.push(parseToken(tokens));
  }
  return result;
}

function parseToken(tokens) {
  if (tokens.length === 0) throw new Error("unexpected EOF");
  const token = tokens.shift();
  if (token === "(") {
    const expr = [];
    while (tokens[0] !== ")") {
      expr.push(parseToken(tokens));
    }
    tokens.shift();
    return expr;
  }
  if (token === ")") throw new Error("unexpected )");
  return categorize(token);
}

function categorize(token) {
  if (!isNaN(parseFloat(token)))
    return { type: "literal", value: parseFloat(token) };
  // a string
  if (token[0] === '"' && token.slice(-1) === '"')
    return { type: "literal", value: token.slice(1, -1) };
  return { type: "identifier", value: token };
}

export { parse };

File: python/parse.py
from typing import List
from python.err import ParseError
from python.token import Token, TokenType
from python.utils import is_number


def Tokenize(input):
    return input.replace('(', ' ( ').replace(')', ' ) ').split()


def ParseTokens(tokens):
    if len(tokens) == 0:
        return None
    token = tokens.pop(0)
    if token == '(':
        exprs = []
        while tokens[0] != ')':
            exprs.append(ParseTokens(tokens))
        tokens.pop(0)
        return exprs

    if token == ')':
        raise ParseError

    return token


def categorize(token):
    if isinstance(token, List):
        if token[0] == "\"" and token[-1] == "\"":
            return Token(TokenType.Identifier, token[1:-1])
    if is_number(token):
        return Token(TokenType.Literal, float(token))
    return Token(TokenType.Identifier, token)
